pkg/flow/internal/importsource/import_file.go

```diff
@@ -3,96 +3,236 @@ package importsource
+	Type filedetector.Detector `river:"detector,attr,optional"`
 	// PollFrequency determines the frequency to check for changes when Type is Poll.
 	PollFrequency time.Duration `river:"poll_frequency,attr,optional"`
 }
 
 var DefaultFileArguments = FileArguments{
-	Type:          file.DetectorFSNotify,
+	Type:          filedetector.DetectorFSNotify,
 	PollFrequency: time.Minute,
 }
 
-type importFileConfigBlock struct {
-	LocalFileArguments FileArguments `river:",squash"`
-}
-
 // SetToDefault implements river.Defaulter.
-func (a *importFileConfigBlock) SetToDefault() {
-	a.LocalFileArguments = DefaultFileArguments
+func (a *FileArguments) SetToDefault() {
+	*a = DefaultFileArguments
 }
 
 func (im *ImportFile) Evaluate(scope *vm.Scope) error {
-	var arguments importFileConfigBlock
+	var arguments FileArguments
 	if err := im.eval.Evaluate(scope, &arguments); err != nil {
 		return fmt.Errorf("decoding River: %w", err)
 	}
-	if im.fileComponent == nil {
-		var err error
-		im.fileComponent, err = file.New(im.managedOpts, file.Arguments{
-			Filename:      arguments.LocalFileArguments.Filename,
-			Type:          arguments.LocalFileArguments.Type,
-			PollFrequency: arguments.LocalFileArguments.PollFrequency,
-			// isSecret is only used for exported values; modules are not exported
-			IsSecret: false,
-		})
-		if err != nil {
-			return fmt.Errorf("creating file component: %w", err)
-		}
-		im.arguments = arguments
-	}
 
-	if reflect.DeepEqual(im.arguments, arguments) {
+	if reflect.DeepEqual(im.args, arguments) {
 		return nil
 	}
+	im.args = arguments
 
-	// Update the existing managed component
-	if err := im.fileComponent.Update(arguments); err != nil {
-		return fmt.Errorf("updating component: %w", err)
+	// Force an immediate read of the file to report any potential errors early.
+	if err := im.readFile(); err != nil {
+		return fmt.Errorf("failed to read file: %w", err)
 	}
-	im.arguments = arguments
-	return nil
+
+	reloadFile := func() {
+		select {
+		case im.reloadCh <- struct{}{}:
+		default:
+			// no-op: a reload is already queued so we don't need to queue a second
+			// one.
+		}
+	}
+
+	var err error
+	switch im.args.Type {
+	case filedetector.DetectorPoll:
+		im.detector = filedetector.NewPoller(filedetector.PollerOptions{
+			Filename:      im.args.Filename,
+			ReloadFile:    reloadFile,
+			PollFrequency: im.args.PollFrequency,
+		})
+	case filedetector.DetectorFSNotify:
+		im.detector, err = filedetector.NewFSNotify(filedetector.FsNotifyOptions{
+			Logger:        im.managedOpts.Logger,
+			Filename:      im.args.Filename,
+			ReloadFile:    reloadFile,
+			PollFrequency: im.args.PollFrequency,
+		})
+	}
+
+	return err
 }
 
 func (im *ImportFile) Run(ctx context.Context) error {
-	return im.fileComponent.Run(ctx)
+	defer func() {
+		if err := im.detector.Close(); err != nil {
+			level.Error(im.managedOpts.Logger).Log("msg", "failed to shut down detector", "err", err)
+		}
+		im.detector = nil
```

Comment:
`im.detector` is protected by a mutex, so you should grab the mutex here before closing and resetting it